### PR TITLE
Accept both singular/multiple values in applicable functions in the Python client

### LINKED ISSUE #3558: Support Union[str, List[str]) for table operations in the Python client to be consistent with the server API

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #3558

### CODE REVIEW

File: py/client/pydeephaven/_table_interface.py [hunk 1]
```diff
@@ -195,56 +197,57 @@ def tail(self, num_rows: int):
...
 
         Args:
             table (Table): the right-table of the join
-            on (List[str]): the columns to match, can be a common name or an equal expression,
+            on (Union[str, List[str]]): the column(s) to match, can be a common name or an equal expression,
                 i.e. "col_a = col_b" for different column names
-            joins (List[str], optional): a list of the columns to be added from the right table to the result
-                table, can be renaming expressions, i.e. "new_col = col"; default is empty
+            joins (Union[str, List[str]], optional): the column(s) to be added from the right table to the result
+                table, can be renaming expressions, i.e. "new_col = col"; default is None
```
Comment (chipkent) [inline py/client/pydeephaven/_table_interface.py:209]: On the `joins` for the join methods, we are not clear about what the default does.

File: py/client/pydeephaven/_table_interface.py [hunk 2]
```diff
@@ -195,56 +197,57 @@ def tail(self, num_rows: int):
...
         returns the result table.
 
         Args:
             table (Table): the right-table of the join
-            on (List[str]): the columns to match, can be a common name or an equal expression,
-                i.e. "col_a = col_b" for different column names
-            joins (List[str], optional): a list of the columns to be added from the right table to the result
-                table, can be renaming expressions, i.e. "new_col = col"; default is empty
+            on (Union[str, List[str]]): the column(s) to match, can be a common name or an equal expression,
+                i.e. "col_a = col_b" for different column names, default is None
```
Comment (chipkent) [inline py/client/pydeephaven/_table_interface.py:248]: Not clear what the default does.

File: py/client/pydeephaven/_table_interface.py [hunk 3]
```diff
@@ -367,192 +371,193 @@ def ungroup(self, cols: List[str] = [], null_fill: bool = True):
...
         return self.table_op_handler(table_op)
 
-    def first_by(self, by: List[str] = []):
+    def first_by(self, by: Union[str, List[str]] = None):
         """ Perform First-by aggregation on the table and return the result table which contains the first row of each
         distinct group.
 
         Args:
-            by (List[str], optional): the group-by column names, default is empty
+            by (Union[str, List[str]], optional): the group-by column name(s), default is None
```
Comment (chipkent) [inline py/client/pydeephaven/_table_interface.py:382]: On these by methods we do not explain what the default does.

File: py/client/pydeephaven/query.py [hunk 4]
```diff
@@ -163,285 +165,276 @@ def tail(self, num_rows: int):
...
 
         Args:
             table (Table): the right-table of the join
-            on (List[str]): the columns to match, can be a common name or an equal expression,
+            on (Union[str, List[str]]): the column(s) to match, can be a common name or an equal expression,
                 i.e. "col_a = col_b" for different column names
-            joins (List[str], optional): a list of the columns to be added from the right table to the result
-                table, can be renaming expressions, i.e. "new_col = col"; default is empty
+            joins (Union[str, List[str]], optional): the column(s) to be added from the right table to the result
+                table, can be renaming expressions, i.e. "new_col = col"; default is None
```
Comment (chipkent) [inline py/client/pydeephaven/query.py:207]: See the other comments about us not explaining defaults.  There are many more in this file.

File: py/client/tests/test_session.py [hunk 5]
```diff
@@ -221,7 +221,7 @@ def test_input_table(self):
         dh_table = self.session.import_table(pa_table)
```
Comment (chipkent) [inline py/client/tests/test_session.py:221]: Should we unit test any of the other cases that were changed?  I don't have a strong view, and you could easily argue that "no" is the right answer.

**Review by chipkent — CHANGES_REQUESTED** [review]

File: py/client/tests/test_session.py (hunk 5 above)
Comment (jmao-denver) [reply py/client/tests/test_session.py:221]: Yeah, that would be a over-kill and bad ROI

**Review by chipkent — APPROVED** [review]

**Pull request merged**